MaxText/layers/linears.py

```diff
@@ -436,25 +436,25 @@ def generate_masks(self, top_k_indices, softmax_probs):
       # so the 2nd token for expert #1 ([0, 1] & [1, 3]) is dropped, output of updated_expert_mask is [[[1, 1],[0, 1]]].
       expert_mask = jax.nn.one_hot(top_k_indices, num_classes=self.num_experts, dtype=jnp.int32)
       expert_mask_fused = jnp.reshape(expert_mask, (batch_size, seq_len * self.num_experts_per_tok, self.num_experts))
+      expert_mask_fused = nn.with_logical_constraint(expert_mask_fused, ("activation_batch", None, None))
       expert_token_count_fused = jnp.cumsum(expert_mask_fused, axis=1)
       expert_token_count = jnp.reshape(expert_token_count_fused, ((batch_size, seq_len, self.num_experts_per_tok, self.num_experts)))
+      expert_token_count = nn.with_logical_constraint(expert_token_count, ("activation_batch", "activation_length", None, None))
       trunc_expert_mask = expert_mask * jnp.less_equal(expert_token_count, expert_capacity_per_batch)
-      updated_expert_mask = jnp.sum(trunc_expert_mask, axis=3)
+      combined_expert_mask = jnp.sum(trunc_expert_mask, axis=2)
 
       # reshape & update weights
-      softmax_probs *= updated_expert_mask
-      weights = self.reshape_and_update_weights(softmax_probs, top_k_indices)
```

Comment:
We may have already discussed this - are we able to drop this here since its mostly a no-op (previous convo https://screenshot.googleplex.com/6mFBfbdjeXGCfSy?)

Reply:
Yes, it's safe to drop here with changing 

`updated_expert_mask = jnp.sum(trunc_expert_mask, axis=3)` to `combined_expert_mask = jnp.sum(trunc_expert_mask, axis=2)` at line #442. I also tested in the moe unit test, the output is the same. However, this reshape_and_update_weights function is still needed for non-dropping implementation